driver/main.c

```diff
@@ -1275,8 +1276,49 @@ static inline void record_drop_x(struct ppm_consumer_t *consumer, struct timespe
 	}
 }
 
-static inline int drop_event(struct ppm_consumer_t *consumer, enum ppm_event_type event_type, enum syscall_flags drop_flags, struct timespec *ts)
+static inline int drop_event(struct ppm_consumer_t *consumer,
+			     enum ppm_event_type event_type,
+			     enum syscall_flags drop_flags,
+			     struct timespec *ts,
+			     struct pt_regs *regs)
 {
+	unsigned long close_fd = -1;
+	struct files_struct *files;
+	struct fdtable *fdt;
+	bool close_return = false;
+
+	// It's annoying but valid for a program to make a large number of
+	// close() calls on nonexistent fds. That can cause driver cpu usage
+	// to spike dramatically, so drop close events if the fd is not valid.
+	if (consumer->dropping_mode) {
+		if (event_type == PPME_SYSCALL_CLOSE_X) {
+			if (syscall_get_return_value(current, regs) < 0) {
+				close_return = true;
+			}
+		} else if (event_type == PPME_SYSCALL_CLOSE_E) {
+			syscall_get_arguments(current, regs, 0, 1, &close_fd);
+
+			files = current->files;
+			spin_lock(&files->file_lock);
+			fdt = files_fdtable(files);
+			if (close_fd >= fdt->max_fds) {
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(3, 4, 0))
+				if (!FD_ISSET(close_fd, fdt->open_fds))
+#else
+				if (!fd_is_open(close_fd, fdt))
+#endif
+				{
+					close_return = true;
+				}
+			}
+			spin_unlock(&files->file_lock);
+		}
+
+		if (close_return) {
+			return 1;
+		}
+	}
+
 	if (drop_flags & UF_NEVER_DROP) {
```

Comment:
Is the added code placed before this check on purpose? This means even when `drop_flags` is UF_NEVER_DROP, some close events will be dropped.